Adventure 5.0

kezz opened the pull request:
### Introduction
This PR exists to track the progress on Adventure 5.0. It only exists as a PR for review purposes, once Adventure 5.0 is complete, the main branch of the repo will be changed to [`main/5`](https://github.com/KyoriPowered/adventure/tree/main/5).

Discussion in this PR will be limited to reviews of the PR itself. For discussion about broader changes in Adventure 5.0, please see https://github.com/KyoriPowered/adventure/issues/1202.

### To-do list:
The list below tracks remaining tasks. For reference, "refactor" means removal of all deprecations and refactoring of code to use modern Java language features.

- [x] Bump Java version to 21.
- [x] Refactor `annotation-processors` module.
- [x] Refactor `api` module.
- [x] Remove `extra-kotlin` module.
- [x] Refactor `key` module.
- [x] Refactor `nbt` module.
- [x] Refactor `serializer-configurate-4` module.
- [x] Refactor `text-logger-slf4j` module.
- [x] Refactor `text-minimessage` module.
- [x] Refactor `text-serializer-ansi` module.
- [x] Refactor `text-serializer-commons` module.
- [x] Refactor `text-serializer-gson` module.
- [x] Refactor `text-serializer-gson-legacy-impl` module.
- [x] Refactor `text-serializer-json` module.
- [x] Refactor `text-serializer-json-legacy` module.
- [x] Refactor `text-serializer-legacy` module.
- [x] Refactor `text-serializer-plain` module.
- [x] Fix GitHub actions builds.
- [x] Check all Javadocs, IJ's reformatting changes JD references to impl classes and fucks up spacing for no reason.
- [x] Migrate to jspecify.
- [x] Double check jspecify everywhere, some weirdness remaining.
- [x] Make `TextFormat` sealed.
- [x] Modularize!
- [x] Find some sort of abi/api break checker just to double check we've not removed anything important.
- [X] Check if there's any libraries we can/should update
- [x] Add a note on Component, Payload, Action that new types may be added at any time

### 4.0 changes required
Deprecate:
- [x] BuildableComponent
- [X] The "enumness" of ClickEvent$Action
- [X] ClickEvent$Action#payloadType

Add `ApiStatus.NonExtendible` to:
- [x] Title.Times
- [x] Sound.Builder
- [x] ResourcePackRequest (and Builder)
- [x] ResourcePackInfo (and Builder)
- [x] PointersSupplier (and Builder)
- [x] Pointers (and Builder)
- [x] Pointer
- [x] Identity
- [x] ComponentFlattener (and Builder)
- [x] Basically everything in adventure-nbt

### Remaining discussion points
- [x] Do we bump to Java 25 upon release?
_i dont think so, doesn't seem like mc is bumping any time soon?_
- [X] Look at other features targeted for 5.0 release or tagged with "needs discussion".
- [x] Do we want to type click event? **yes**
- [x] Decide if we want Examinable to be public API or not **yeeted completely**
       _atm it's literally just used for toString, but is sometimes in the API classes and sometimes impl only - we should be consistent here_
- [X] Do we want to yeet all component constructors to be simply `data`, `data, style`, `data, sba...`? **yes, with old methods left as synthetic, actually no now - will push to 6.x**

kezz committed:
build: Bump Java version to 21, remove extra-kotlin module

kezz committed:
chore(key): Refactor module to remove deprecations and use modern language features

kezz committed:
chore(text-logger-slf4j): Refactor module to use modern language features

Also adds a missing `PlatformAPI` annotation and removes an unused method in CallerClassFinder.

kezz committed:
chore(api): Initial refactor (minus `text` package) to remove deprecations and use modern language features

jpenilla committed:
Update indra to 4.0.0-SNAPSHOT

Strokkur424 committed:
chore(nbt): modernize and remove deprecated methods

Strokkur424 committed:
chore(configurate-4): cleanup code

Strokkur424 committed:
chore(minimessage): modernize some code

Strokkur424 committed:
chore(ansi): replace one switch expression

Strokkur424 committed:
chore(serializer-gson): modernize gson serializer and remove legacy impl

Strokkur424 committed:
chore(serializer-json): cleanup and modernize json modules

Strokkur424 committed:
chore(serializer-legacy): cleanup and modernize code

Strokkur424 committed:
chore(serializer-plain): remove deprecated classes

Strokkur424 committed:
feat: add module-info.java to all modules

kezz committed:
chore(api): Further refactor (including `text` package) to remove deprecations and use modern language features

Strokkur424 committed:
chore: migrate to jspecify

kezz committed:
chore(key): Remove examination and fix module exports

kezz committed:
chore(nbt): Remove examination, fix module exports and record-ify everything

kezz committed:
chore(api): Remove examination, fix module exports and a few other minor fixups

kashike commented on api/src/main/java/module-info.java:
```diff
@@ -0,0 +1,38 @@
+@NullMarked
+module net.kyori.adventure.api {
+  requires transitive net.kyori.adventure.key;
+  requires transitive static org.jspecify;
```
See: https://github.com/jspecify/jspecify/issues/495

kezz committed:
chore(api): Fix checkstyle, build and further misc refactoring

kezz committed:
chore(nbt): Fix checkstyle, build and further misc refactoring

kezz committed:
chore(misc): Continue work on checkstyle, build, and misc refactoring

kezz committed:
chore(deps): Bump gradle to 9.1.0

Fixes an error when building.

kezz committed:
chore(nbt): Disable pretty printing test

zml2008 committed:
chore(build): Run GHA on J21

Also revert build-logic project to J17, to align with Gradle's minimum.

kezz committed:
fix(api): Remove incorrect BossBar$Listener annotations

kezz committed:
chore(api): Update since + nbt return type for toBuilder

kermandev commented on key/src/main/java/net/kyori/adventure/key/Key.java:
```diff
@@ -135,7 +132,9 @@ public interface Key extends Comparable<Key>, Examinable, Namespaced, Keyed {
    */
-  static @NotNull Key key(@KeyPattern.Namespace final @NotNull String namespace, @KeyPattern.Value final @NotNull String value) {
+  static Key key(@KeyPattern.Namespace final String namespace, @KeyPattern.Value final String value) {
+    KeyImpl.checkError("namespace", namespace, namespace, value, Key.checkNamespace(namespace), KeyImpl.NAMESPACE_PATTERN);
```
Any major reason this isnt moved to the canonical constructor of KeyImpl? Seems like its a common pattern here and could be over sight, ex. KeyedValueImpl can actually accept null keys.

The only reason I could think of is providing a trusted namespace/value object.

kezz committed:
fix(key): Move null/format checks into canonical constructors

kezz committed:
feature(api): Use sealed subclass system for click event action

kezz committed:
docs(api): Add notes about sealed interfaces that may be added to at any time

zml2008 committed:
chore(deps): indra 4 release

zml2008 committed:
Merge pull request #1289 from KyoriPowered/indra-4

Update indra to v4

zml2008 committed:
chore(api): resolve some of the warnings and test compile errors

kezz committed:
fix(api): Restore missing score component with obsolete annotation, add new obsoleteSinceMinecraft jd param

kezz committed:
chore(api): Fix checkstyle errors in ClickEventImpl

kezz committed:
feature(minimessage): Migrate to modern language features and fix build (mostly)

kezz committed:
chore(minimessage): Fix more checkstyle errors

zml2008 committed:
feat!(text-logger-slf4j): Require slf4j2

This reverts commit 5a2ebc88ea96d1ea284179437ae7d51a465781b8.

A hard dependency on slf4j2 allows us to provide wrappers around the fluent API.

kezz committed:
chore(api,legacy): Seal TextFormat and fix #1328

Also fixes some build/checkstyle issues

zml2008 committed:
chore(ci): Update for org move

kezz committed:
chore(key): Fix remaining build warnings and move null-marking to package-level

Strokkur424 committed:
fix: replace all module-level NullMarked with package-level ones

Strokkur424 committed:
Merge branch 'main/5' into chore/fix-nullability

Strokkur424 committed:
chore: annotated net.kyori.test package as NullMarked

Strokkur424 committed:
fix(api): start fixing api tests

kermandev committed:
fix(api): Unbound profileProperties Collection

This should allow subtypes of ProfileProperty, which is commonly used by the
implementing platform, to return a different type that may contain more information.

This is not an ABI break.

Strokkur424 committed:
fix(api): score is no longer flattened, tests should reflect that

Strokkur424 committed:
fix: a bunch of compilation errors and failing tests

Strokkur424 committed:
chore: bump gson version and fix checkstyle errors

Strokkur424 committed:
chore: minor changes

Strokkur424 committed:
fix: set interpret default to false

kezz committed:
Merge pull request #1333 from Strokkur424/chore/fix-nullability

chore: my additions to main/5

kezz committed:
chore(api): Add scoped component toBuilder override and add some no-args constructors

kezz committed:
fix(serializer-gson,serializer-configurate4): Fix remaining compile errors

Strokkur424 committed:
chore: fix remaining tests

kezz committed:
Merge pull request #1343 from Strokkur424/fix/tests

chore: fix remaining tests

kezz committed:
chore(api): Restore NamedTextColor#toString behavior

kezz committed:
chore(api): Restore ClickEvent.Action#name

renovate[bot] committed:
chore(deps): Update guava monorepo to v33.5.0-jre

renovate[bot] committed:
chore(deps): Update dependency org.junit:junit-bom to v6

kezz committed:
feature(api): Make custom payload NBT optional

kezz committed:
feature(minimessage,serializer-gson): Correctly handle nullable custom payloads

kezz committed:
chore: Migrate to Java 9+ immutable collection methods

Closes #459.

renovate[bot] committed:
chore(deps): Update dependency com.puppycrawl.tools:checkstyle to v12

kezz committed:
Merge pull request #1345 from PaperMC/feature/optional-nbt-payload

Optional NBT payload for custom click events

kezz committed:
Merge pull request #1342 from PaperMC/renovate/major-junit-framework-monorepo

chore(deps): Update dependency org.junit:junit-bom to v6

kezz committed:
Merge pull request #1341 from PaperMC/renovate/major-checkstyle

chore(deps): Update dependency com.puppycrawl.tools:checkstyle to v12

kezz committed:
Merge pull request #1340 from PaperMC/renovate/guava-monorepo

chore(deps): Update guava monorepo to v33.5.0-jre

kezz committed:
Merge pull request #1337 from kermandev/fix/playerobject-bounds

fix(api): Unbound profileProperties Collection

kezz committed:
feature(api): Add BookLike interface

kezz committed:
fix(api): Fix since tag for custom click events

kezz committed:
Merge pull request #1346 from PaperMC/feature/book-like

feature(api): Add BookLike interface

kezz committed:
fix(api): Unseal SignedMessage

Platforms can implement this with actual signed messages (not just system ones), so it needs to be unsealed.

kezz committed:
feature(api): Make ChatType#key nullable, and remove Keyed implementation

kezz committed:
chore(api): Move more deprecations to obsolete annotations

kezz committed:
chore(api): Refactor Index from interface to class

kezz committed:
chore(api): Refactor NamedTextColor from interface to class

kezz committed:
chore(api): Refactor HoverEvent (and subclasses) from interface to class

kezz committed:
chore(api): Refactor ClickEvent and Action from interface to class

kezz committed:
chore(api): Add missing style buildable extends

kezz committed:
chore(api): Remove UTF8ResourceBundleControl

Resource bundles are UTF-8 by default since Java 9.

renovate[bot] committed:
chore(deps): Update dependency com.diffplug.gradle:goomph to v4.4.1

renovate[bot] committed:
chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.45.0

renovate[bot] committed:
chore(deps): Update gradle and github actions

zml2008 committed:
bump ci-cookbook

zml2008 committed:
Merge pull request #1258 from PaperMC/renovate/gradle-and-github-actions

chore(deps): Update gradle and github actions

zml2008 committed:
Merge pull request #1339 from PaperMC/renovate/error-prone-monorepo

chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.45.0

zml2008 committed:
Merge pull request #1338 from PaperMC/renovate/patch-release-dependency-changes

chore(deps): Update dependency com.diffplug.gradle:goomph to v4.4.1

renovate[bot] committed:
chore(deps): Update dependency com.puppycrawl.tools:checkstyle to v12.3.0

zml2008 committed:
Merge pull request #1352 from PaperMC/renovate/checkstyle

chore(deps): Update dependency com.puppycrawl.tools:checkstyle to v12.3.0

MiniDigger committed:
chore: bump ci-cookbook

zml2008 committed:
this should be a tag

RabbitTV22 committed:
add femboy flag

RabbitTV22 committed:
Merge branch 'PaperMC:main/5' into main/5

zml2008 committed:
Merge pull request #1357 from RabbitTV22/main/5

Add femboy flag to pride tag

renovate[bot] committed:
chore(deps): Update patch-release dependency changes

zml2008 committed:
Merge pull request #1355 from PaperMC/renovate/patch-release-dependency-changes

chore(deps): Update patch-release dependency changes

renovate[bot] committed:
chore(deps): Update Gradle to v9.3.1

renovate[bot] committed:
chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.47.0

zml2008 committed:
Merge pull request #1363 from PaperMC/renovate/gradle-and-github-actions

chore(deps): Update Gradle to v9.3.1

zml2008 committed:
chore(deps): Bump gradle errorprone plugin version

renovate[bot] committed:
chore(deps): Update KyoriPowered/ci-cookbook action to v2026

zml2008 committed:
Merge pull request #1365 from PaperMC/renovate/major-gradle-and-github-actions

chore(deps): Update KyoriPowered/ci-cookbook action to v2026

zml2008 committed:
Merge pull request #1362 from PaperMC/renovate/error-prone-monorepo

chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.47.0

Privatech38 committed:
feat: Add toString for StyleImpl

Emilxyz committed:
chore: add back dropped toString implementations

Emilxyz committed:
fix(text-minimessage): support all uuid versions in head tag

kezz committed:
Merge pull request #1375 from Emilxyz/fix/head-tag-uuid

fix(text-minimessage): support all uuid versions in head tag

kezz committed:
Merge pull request #1374 from Emilxyz/chore/dropped-tostring

chore: add back dropped toString implementations

kezz committed:
Merge pull request #1373 from Privatech38/feat/style-impl-to-string

feat(api): Add toString implementation for StyleImpl

renovate[bot] committed:
chore(deps): Update dependency org.junit:junit-bom to v6.0.3

renovate[bot] committed:
chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.48.0

renovate[bot] committed:
chore(deps): Update dependency org.jetbrains:annotations to v26.1.0

zml2008 committed:
Merge pull request #1380 from PaperMC/renovate/patch-release-dependency-changes

chore(deps): Update dependency org.junit:junit-bom to v6.0.3

zml2008 committed:
Merge pull request #1381 from PaperMC/renovate/error-prone-monorepo

chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.48.0

zml2008 committed:
Merge pull request #1382 from PaperMC/renovate/org.jetbrains-annotations-26.x

chore(deps): Update dependency org.jetbrains:annotations to v26.1.0

kezz committed:
chore: Fix a bunch more build warnings

kezz commented:
Barring the remaining three PRs in the 5.0 milestone, 5.0 is now considered complete. I'll close this PR now, but any more issues can be made directly on the repo!
